revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,246 @@
+		ServerResults: []*result.ServerResult{{
+			Result: result.ResultNonRevokable,
+			Error:  nil,
+		}},
+	}
+
+	wg.Wait()
+	return certResults, nil
+}
+
+func certCheckStatus(cert, issuer *x509.Certificate, opts Options) *result.CertRevocationResult {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return &result.CertRevocationResult{
+			Result:        result.ResultNonRevokable,
+			ServerResults: []*result.ServerResult{errToServerResult("", NoOCSPServerError{})},
+		}
+	}
+
+	serverResults := make([]*result.ServerResult, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		serverResult := checkStatusFromServer(cert, issuer, server, opts)
+		if serverResult.Result == result.ResultOK || serverResult.Result == result.ResultRevoked || errors.Is(serverResult.Error, UnknownStatusError{}) {
+			// A valid response has been received from an OCSP server
+			// Result should be based on only this response, not any errors from
+			// other servers
+			return serverResultsToCertRevocationResult([]*result.ServerResult{serverResult})
+		}
+		serverResults[serverIndex] = serverResult
+	}
+	return serverResultsToCertRevocationResult(serverResults)
+}
+
+func checkStatusFromServer(cert, issuer *x509.Certificate, server string, opts Options) *result.ServerResult {
+	// Check valid server
+	if serverURL, err := url.Parse(server); err != nil || strings.ToLower(serverURL.Scheme) != "http" {
+		// This function is only able to check servers that are accessible via HTTP
+		return errToServerResult(server, OCSPCheckError{Err: fmt.Errorf("OCSPServer protocol %s is not supported", serverURL.Scheme)})
+	}
+
+	// Create OCSP Request
+	resp, err := executeOCSPCheck(cert, issuer, server, opts)
+	if err != nil {
+		// If there is a server error, attempt all servers before determining what to return
+		// to the user
+		return errToServerResult(server, err)
+	}
+
+	// Validate OCSP response isn't expired
+	if time.Now().After(resp.NextUpdate) {
+		return errToServerResult(server, OCSPCheckError{Err: errors.New("expired OCSP response")})
+	}
+
+	// Check for presence of pkix-ocsp-no-check extension in response
+	foundNoCheck := false
+	for _, extension := range resp.Extensions {
+		if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+			foundNoCheck = true
+			break
+		}
+	}
+	//lint:ignore SA9003 // Empty body in an if branch (Remove after adding CRL)
+	if !foundNoCheck {
+		// This will be ignored until CRL is implemented
+		// If it isn't found, CRL should be used to verify the OCSP response
+		// TODO: add CRL support
+		// https://github.com/notaryproject/notation-core-go/issues/125
+	}
+
+	// No errors, valid server response
+	switch resp.Status {
+	case ocsp.Good:
+		return errToServerResult(server, nil)
+	case ocsp.Revoked:
+		// Check if SigningTime was before the revocation
+		if opts.SigningTime.Before(resp.RevokedAt) {
+			return errToServerResult(server, nil)
+		}
```

Comment:
Reminder for me: Need to verify .

Reply:
As per [rfc6960#section-4.4.5](https://www.rfc-editor.org/rfc/rfc6960#section-4.4.5) ocsp can contain all CRL extension so as per RFC it should be [invalidtiyDate](https://www.rfc-editor.org/rfc/rfc5280#section-5.3.2).
> ```All the extensions specified as CRL entry extensions -- in [Section 5.3 of [RFC5280]](https://www.rfc-editor.org/rfc/rfc5280#section-5.3) -- are also supported as singleExtensions```

Also as per [CA/B forum BR](https://github.com/cabforum/code-signing/blame/main/docs/CSBR.md#L555) 
> ```Note: Backdating the revocationDate field is an exception to best practice described in RFC 5280 (section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.```

Since notation can be used with private CA, the CA/B forum guideline might not work for us. Going to do some more reading.

Reply:
Although invalidityDate is supported, I don't believe this is the preference for OCSP. Single extensions would mean it is supported as an extra piece of information. This is the description of invalidityDate within RFC5280 for CRL:
> The invalidity date is a non-critical CRL entry extension that provides the date on which it is known or suspected that the private key was compromised or that the certificate otherwise became invalid.

Similarly, here is the description of revocationTime from RFC6960 for OCSP:
> Responses defined in this document can contain four times -- thisUpdate, nextUpdate, producedAt, and revocationTime.  The semantics of these fields are:
> ...
>    revocationTime: The time at which the certificate was revoked or placed on hold.

Since they are so similar, but revocationTime is defined specifically for OCSP, I think it makes the most sense to use this value (especially because it is not optional)

Reply:
Asked for suggestions in slack https://cloud-native.slack.com/archives/CQUH8U287/p1681778707365059

Reply:
Resolving. Implementation will follow current spec for RC-4. If change in behavior is agreed on Monday sync, spec will be updated and implementation will be updated to follow check in a new PR.